src/i_joystick.c

```diff
@@ -76,6 +81,246 @@ static int joystick_physical_buttons[NUM_VIRTUAL_BUTTONS] = {
+
+        // First, look for the gamepad at the previously-used index.
+        if (joystick_index >= 0 && joystick_index < SDL_NumJoysticks())
+        {
+            dev_guid = SDL_JoystickGetDeviceGUID(joystick_index);
+            if (!memcmp(&guid, &dev_guid, sizeof(SDL_JoystickGUID)))
+            {
+                return joystick_index;
+            }
+        }
+
+        // Maybe the index has moved?
+        index = FindSpecificGamepad(guid);
+    }
+
+    // Find a gamepad for the first time. Or, if the previous gamepad isn't
+    // present, see if a different one is available.
+    if (index < 0)
+    {
+        index = FindFirstGamepad();
+    }
+
+    return index;
+}
+
+void I_InitGamepad(void)
+{
+    SDL_JoystickGUID guid;
+    int index;
+
+    if (!use_gamepad)
+    {
+        return;
+    }
+
+    if (SDL_Init(SDL_INIT_GAMECONTROLLER) < 0)
+    {
+        return;
+    }
+
+    index = DeviceIndexGamepad();
+
+    if (index < 0)
+    {
+        printf("I_InitGamepad: No gamepad found.\n");
+        SDL_QuitSubSystem(SDL_INIT_GAMECONTROLLER);
+        return;
+    }
+
+    gamepad = SDL_GameControllerOpen(index);
+
+    if (gamepad == NULL)
+    {
+        printf("I_InitGamepad: Failed to open gamepad.\n");
+        SDL_QuitSubSystem(SDL_INIT_GAMECONTROLLER);
+        return;
+    }
+
+    joystick_index = index;
+
+    if (strcmp(joystick_guid, ""))
+    {
+        joystick_guid = malloc(33);
+    }
+
+    guid = SDL_JoystickGetDeviceGUID(joystick_index);
+    SDL_JoystickGetGUIDString(guid, joystick_guid, 33);
+
+    // GameController events do not fire if Joystick events are disabled.
+    SDL_JoystickEventState(SDL_ENABLE);
+    SDL_GameControllerEventState(SDL_ENABLE);
+
+    printf("I_InitGamepad: %s\n", SDL_GameControllerName(gamepad));
+    I_AtExit(I_ShutdownGamepad, true);
+}
+
+static int GetTriggerStateGamepad(SDL_GameControllerAxis trigger)
+{
+    int value;
+
+    value = SDL_GameControllerGetAxis(gamepad, trigger);
+
+    return (value > TRIGGER_THRESHOLD);
+}
+
+// Get the state of the given virtual button.
+
+static int ReadButtonStateGamepad(int vbutton)
+{
+    int physbutton, state;
+
+    // Map from virtual button to physical (SDL) button.
+    if (vbutton < NUM_VIRTUAL_BUTTONS)
+    {
+        physbutton = joystick_physical_buttons[vbutton];
+    }
+    else
+    {
+        physbutton = vbutton;
+    }
+
+    switch (physbutton)
+    {
+        case GAMEPAD_BUTTON_TRIGGERLEFT:
+            state = GetTriggerStateGamepad(SDL_CONTROLLER_AXIS_TRIGGERLEFT);
+            break;
+
+        case GAMEPAD_BUTTON_TRIGGERRIGHT:
+            state = GetTriggerStateGamepad(SDL_CONTROLLER_AXIS_TRIGGERRIGHT);
+            break;
+
+        default:
+            state = SDL_GameControllerGetButton(gamepad, physbutton);
+            break;
+    }
+
+    return state;
+}
+
+// Get a bitmask of all currently-pressed buttons
+
+static int GetButtonsStateGamepad(void)
+{
+    int i;
+    int result;
+
+    result = 0;
```

Comment:
Make this part of the variable declaration.

Reply:
0e808551